operators/src/processing/vector_join/equi_data_join.rs

```diff
@@ -0,0 +1,401 @@
+                Self::compute_join(&c1, c2, left, &right)
+            }
+            (FeatureDataRef::Decimal(c1), FeatureDataRef::Decimal(c2)) => {
+                Self::compute_join(&c1, c2, left, &right)
+            }
+            (FeatureDataRef::Text(c1), FeatureDataRef::Text(c2)) => {
+                Self::compute_join(&c1, c2, left, &right)
+            }
+            (FeatureDataRef::Categorical(c1), FeatureDataRef::Categorical(c2)) => {
+                Self::compute_join(&c1, c2, left, &right)
+            }
+            (left, right) => Err(Error::ColumnTypeMismatch {
+                left: (&left).into(),
+                right: (&right).into(),
+            }),
+        }
+    }
+
+    fn compute_join<'i, D1, D2, T>(
+        left_join_values: &D1,
+        right_join_values: D2,
+        left: &FeatureCollection<G>,
+        right: &DataCollection,
+    ) -> Result<FeatureCollection<G>>
+    where
+        D1: DataRef<'i, T> + 'i,
+        D2: DataRef<'i, T> + 'i,
+        T: 'i + PartialEq,
+    {
+        let matches_iter = left_join_values
+            .as_ref()
+            .iter()
+            .zip(left.time_intervals())
+            .map(move |(left_value, left_time_interval)| {
+                right_join_values
+                    .as_ref()
+                    .iter()
+                    .zip(right.time_intervals())
+                    .map(|(right_value, right_time_interval)| {
+                        (left_value == right_value)
+                            && left_time_interval.intersects(right_time_interval)
+                    })
+                    .collect::<Vec<bool>>()
+            });
+
+        let mut builder = FeatureCollection::<G>::builder();
+
+        // create header by combining values from both collections
+        for (column_name, column_type) in left
+            .column_types()
+            .into_iter()
+            .chain(right.column_types().into_iter())
+        {
+            builder.add_column(column_name, column_type)?;
+        }
+
+        let mut builder = builder.finish_header();
+
+        for ((left_feature_idx, matches), geometry) in matches_iter
+            .enumerate()
+            .zip(left.geometries().map(Into::into))
+        {
+            // TODO: is it better to traverse column-wise?
+
+            for (right_feature_idx, matched) in matches.into_iter().enumerate() {
+                if !matched {
+                    continue;
+                }
+
+                // add geo
+                builder.push_geometry(geometry.clone())?;
+
+                // add time
+                builder.push_time_interval(
+                    left.time_intervals()[left_feature_idx]
+                        .intersect(&right.time_intervals()[right_feature_idx])
+                        .expect("must intersect for join"),
+                )?;
+
+                // add left values
+                for column_name in left.column_names() {
```

Comment:
call the `column_names` function only once instead of every time.

Reply:
It just iterates over the HashMap index, I don't see any improvement here. It does not create any Vec or other owned type.

Reply:
7d53833